### PR TITLE
ATO-659: JWKS endpoint to publish MFA reset storage token public key

### PR DESCRIPTION
**Pull request by billrdunn**
## What

Create new well-known endpoint on auth frontend API for MFA reset storage token signing public key

## Testing

After deploying to sandpit, key is available at https://auth.sandpit.account.gov.uk/.well-known/mfa-reset-jwk.json

## How to review

Code review only

## Checklist

- [x] Impact on orch and auth mutual dependencies has been checked.

[ATO-661]: https://govukverify.atlassian.net/browse/ATO-661?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: ci/terraform/oidc/api-gateway-frontend.tf [hunk 1]
```diff
@@ -36,6 +36,18 @@ resource "aws_api_gateway_usage_plan_key" "di_auth_frontend_usage_plan_key" {
...
+  parent_id   = aws_api_gateway_rest_api.di_authentication_frontend_api.root_resource_id
+  path_part   = ".well-known"
+}
+
+resource "aws_api_gateway_resource" "auth_frontend_connect_resource" {
+  rest_api_id = aws_api_gateway_rest_api.di_authentication_frontend_api.id
+  parent_id   = aws_api_gateway_rest_api.di_authentication_frontend_api.root_resource_id
+  path_part   = "connect"
+}
+
```
Comment (ethanmills) [inline ci/terraform/oidc/api-gateway-frontend.tf:50]: What is this for?

File: ci/terraform/oidc/mfa-reset-storage-token-jwk.tf [hunk 2]
```diff
@@ -0,0 +1,60 @@
+module "mfa_reset_storage_token_jwk_role" {
+  source      = "../modules/lambda-role"
+  environment = var.environment
+  role_name   = "mfa-reset-jwk-role"
+  vpc_arn     = local.authentication_vpc_arn
+
+  policies_to_attach = [
+    aws_iam_policy.mfa_reset_storage_token_kms_signing_policy.arn
```
Comment (ethanmills) [inline ci/terraform/oidc/mfa-reset-storage-token-jwk.tf:8]: This only needs `GetPublicKey`, not `Sign`

File: ci/terraform/oidc/mfa-reset-storage-token-jwk.tf [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  endpoint_name   = "mfa-reset-jwk.json"
+  path_part       = "mfa-reset-jwk.json"
+  endpoint_method = ["GET"]
+  environment     = var.environment
+
+  handler_environment_variables = {
+    ENVIRONMENT                     = var.environment
+    DOC_APP_TOKEN_SIGNING_KEY_ALIAS = local.doc_app_auth_key_alias_name
+    LOCALSTACK_ENDPOINT             = var.use_localstack ? var.localstack_endpoint : null
```
Comment (ethanmills) [inline ci/terraform/oidc/mfa-reset-storage-token-jwk.tf:23]: Shouldn't need this, use_localstack is always false

File: ci/terraform/oidc/mfa-reset-storage-token-jwk.tf [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    STORAGE_TOKEN_SIGNING_KEY_ALIAS = aws_kms_alias.mfa_reset_signing_key_alias.name
+    HEADERS_CASE_INSENSITIVE        = var.use_localstack ? "true" : "false"
```
Comment (ethanmills) [inline ci/terraform/oidc/mfa-reset-storage-token-jwk.tf:25]: ditto

File: ci/terraform/oidc/mfa-reset-storage-token-jwk.tf [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  security_group_ids                     = [local.authentication_security_group_id]
+  subnet_id                              = local.authentication_private_subnet_ids
+  lambda_role_arn                        = module.mfa_reset_storage_token_jwk_role.arn
+  logging_endpoint_arns                  = var.logging_endpoint_arns
+  cloudwatch_key_arn                     = data.terraform_remote_state.shared.outputs.cloudwatch_encryption_key_arn
+  cloudwatch_log_retention               = var.cloudwatch_log_retention
+  lambda_env_vars_encryption_kms_key_arn = local.lambda_env_vars_encryption_kms_key_arn
+  default_tags                           = local.default_tags
+
+  use_localstack = var.use_localstack
```
Comment (ethanmills) [inline ci/terraform/oidc/mfa-reset-storage-token-jwk.tf:53]: remove or set false

**Comment by ethanmills** [comment]
Happy with this once the dependency on #4639 is met and the comment is removed

File: ci/terraform/oidc/api-gateway-frontend.tf (hunk 1 above)
Comment (billrdunn) [reply ci/terraform/oidc/api-gateway-frontend.tf:50]: I had a problem with the lambda integration in API Gateway which I thought this fixed... On second testing it appears to not be doing anything. Will remove.

Comment (ethanmills) [reply ci/terraform/oidc/api-gateway-frontend.tf]: This is the route for the client registry API in the OIDC API, so definitely shouldn't be needed here!

File: shared/src/main/java/uk/gov/di/authentication/shared/services/JwksService.java [hunk 6]
```diff
@@ -61,6 +61,13 @@ public JWK getPublicDocAppSigningJwkWithOpaqueId() {
         return getPublicJWKWithKeyId(configurationService.getDocAppTokenSigningKeyAlias());
     }
 
+    public JWK getPublicMfaResetStorageTokenJwkWithOpaqueId() {
```
Comment (ethanmills) [inline shared/src/main/java/uk/gov/di/authentication/shared/services/JwksService.java:64]: Is it possible to add a test for this?

File: shared/src/main/java/uk/gov/di/authentication/shared/services/JwksService.java [hunk 7]
```diff
@@ -61,6 +61,13 @@ public JWK getPublicDocAppSigningJwkWithOpaqueId() {
...
+        LOG.info("Retrieving storage token public key");
+        // this is available when ATO-661 is deployed
+        // getPublicJWKWithKeyId(configurationService.getStorageTokenSigningKeyAlias());
+        return getPublicJWKWithKeyId("some-alias");
+    }
```
Comment (ethanmills) [inline shared/src/main/java/uk/gov/di/authentication/shared/services/JwksService.java:69]: Is it possible to add a test for this to appease Sonar?

**Review by ethanmills — APPROVED** [review]
Thanks!

**Pull request merged**